snowflake/src/main/java/org/apache/iceberg/snowflake/entities/SnowflakeIdentifier.java

```diff
@@ -0,0 +1,167 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.snowflake.entities;
+
+import java.util.List;
+import org.apache.commons.dbutils.ResultSetHandler;
+import org.apache.iceberg.relocated.com.google.common.base.Objects;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+
+public class SnowflakeIdentifier {
+  public enum Type {
+    ROOT,
+    DATABASE,
+    SCHEMA,
+    TABLE
+  }
+
+  private String databaseName;
+  private String schemaName;
+  private String tableName;
+
+  protected SnowflakeIdentifier(String databaseName, String schemaName, String tableName) {
+    this.databaseName = databaseName;
+    this.schemaName = schemaName;
+    this.tableName = tableName;
+  }
+
+  public static SnowflakeIdentifier ofRoot() {
+    return new SnowflakeIdentifier(null, null, null);
+  }
+
+  public static SnowflakeIdentifier ofDatabase(String databaseName) {
+    Preconditions.checkArgument(null != databaseName, "databaseName must be non-null");
+    return new SnowflakeIdentifier(databaseName, null, null);
+  }
+
+  public static SnowflakeIdentifier ofSchema(String databaseName, String schemaName) {
+    Preconditions.checkArgument(null != databaseName, "databaseName must be non-null");
+    Preconditions.checkArgument(null != schemaName, "schemaName must be non-null");
+    return new SnowflakeIdentifier(databaseName, schemaName, null);
+  }
+
+  public static SnowflakeIdentifier ofTable(
+      String databaseName, String schemaName, String tableName) {
+    Preconditions.checkArgument(null != databaseName, "databaseName must be non-null");
+    Preconditions.checkArgument(null != schemaName, "schemaName must be non-null");
+    Preconditions.checkArgument(null != tableName, "tableName must be non-null");
+    return new SnowflakeIdentifier(databaseName, schemaName, tableName);
+  }
+
+  /**
+   * If type is TABLE, expect non-null databaseName, schemaName, and tableName. If type is SCHEMA,
+   * expect non-null databaseName and schemaName. If type is DATABASE, expect non-null databaseName.
+   * If type is ROOT, expect all of databaseName, schemaName, and tableName to be null.
+   */
+  public Type getType() {
```

Comment:
If the static factory methods are used, why not just set this as an instance field passed by those methods?

Reply:
Done